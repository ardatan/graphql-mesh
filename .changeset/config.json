{
  "$schema": "https://unpkg.com/@changesets/config@2.1.0/schema.json",
  "commit": false,
  "linked": [],
  "access": "restricted",
  "baseBranch": "master",
  "updateInternalDependencies": "patch",
  "ignore": ["website"],
  "changelog": ["@changesets/changelog-github", { "repo": "urigo/graphql-mesh" }],
  "snapshot": {
    "useCalculatedVersion": true,
    "prereleaseTemplate": "{tag}-{datetime}-{commit}"
  }
}
