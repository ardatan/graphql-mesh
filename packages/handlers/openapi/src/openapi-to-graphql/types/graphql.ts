// Copyright IBM Corp. 2018. All Rights Reserved.
// Node module: openapi-to-graphql
// This file is licensed under the MIT License.
// License text available at https://opensource.org/licenses/MIT

/**
 * Custom type definitions for GraphQL.
 */

import {
  GraphQLObjectType,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLList,
  GraphQLEnumType,
  GraphQLUnionType,
  GraphQLResolveInfo,
} from 'graphql';

export enum GraphQLOperationType {
  Query,
  Mutation,
  // TODO: Subscription
}

export type GraphQLType =
  | GraphQLObjectType
  | GraphQLInputObjectType
  | GraphQLList<any>
  | GraphQLUnionType
  | GraphQLEnumType
  | GraphQLScalarType;

type Arg = {
  type: any;
  description?: string;
};

export type Args = {
  [key: string]: Arg;
};

export type ResolveFunction = (root: any, args: any, ctx: any, info: GraphQLResolveInfo) => Promise<any> | any;

export type Field = {
  type: GraphQLType;
  resolve?: ResolveFunction;
  args?: Args;
  description: string;
};
