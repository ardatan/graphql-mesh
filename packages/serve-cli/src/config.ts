/* eslint-disable @typescript-eslint/no-unused-expressions */
import { lstat } from 'node:fs/promises';
import { isAbsolute, join } from 'node:path';
import { pathToFileURL } from 'node:url';
import type { GatewayConfig, GatewayConfigContext } from '@graphql-mesh/serve-runtime';
import type { KeyValueCache, Logger } from '@graphql-mesh/types';
import type { GatewayCLIBuiltinPluginConfig } from './cli';
import type { ServerConfig } from './server';

export const defaultConfigExtensions = ['.ts', '.mts', '.cts', '.js', '.mjs', '.cjs'];

export const defaultConfigFileName = 'gateway.config';

export function createDefaultConfigPaths(configFileName: string) {
  return defaultConfigExtensions.map(ext => `${configFileName}${ext}`);
}

export async function loadConfig<TContext extends Record<string, any> = Record<string, any>>(opts: {
  quiet?: boolean;
  log: Logger;
  configPath: string | null | undefined;
  configFileName: string;
}) {
  let importedConfig: Partial<
    GatewayConfig<TContext> & ServerConfig & GatewayCLIBuiltinPluginConfig
  > | null = null;

  if (!opts.configPath) {
    !opts.quiet && opts.log.debug(`Searching for default config files`);
    const configPaths = [
      ...createDefaultConfigPaths(defaultConfigFileName),
      ...createDefaultConfigPaths(opts.configFileName),
    ];
    for (const configPath of configPaths) {
      const absoluteConfigPath = join(process.cwd(), configPath);
      const exists = await lstat(absoluteConfigPath)
        .then(() => true)
        .catch(() => false);
      if (exists) {
        !opts.quiet && opts.log.info(`Found default config file ${absoluteConfigPath}`);
        const importUrl = pathToFileURL(absoluteConfigPath).toString();
        const module = await import(importUrl);
        importedConfig = Object(module).gatewayConfig || null;
        if (!importedConfig) {
          !opts.quiet &&
            opts.log.warn(`No "gatewayConfig" exported from config file at ${absoluteConfigPath}`);
        }
        break;
      }
    }
  } else {
    // using user-provided config
    const configPath = isAbsolute(opts.configPath)
      ? opts.configPath
      : join(process.cwd(), opts.configPath);
    !opts.quiet && opts.log.info(`Loading config file at path ${configPath}`);
    const exists = await lstat(configPath)
      .then(() => true)
      .catch(() => false);
    if (!exists) {
      throw new Error(`Cannot find config file at ${configPath}`);
    }
    const importUrl = pathToFileURL(configPath).toString();
    const module = await import(importUrl);
    importedConfig = Object(module).gatewayConfig || null;
    if (!importedConfig) {
      throw new Error(`No "gatewayConfig" exported from config file at ${configPath}`);
    }
  }
  if (importedConfig) {
    !opts.quiet && opts.log.info('Loaded config');
  } else {
    !opts.quiet && opts.log.debug('No config loaded');
  }

  // TODO: validate imported config

  return importedConfig || {};
}

export async function getBuiltinPluginsFromConfig(
  config: GatewayCLIBuiltinPluginConfig,
  ctx: { cache: KeyValueCache },
) {
  const plugins = [];
  if (config.jwt) {
    const { useJWT } = await import('@graphql-mesh/plugin-jwt-auth');
    plugins.push(useJWT(config.jwt));
  }
  if (config.prometheus) {
    const { default: useMeshPrometheus } = await import('@graphql-mesh/plugin-prometheus');
    plugins.push(useMeshPrometheus(config.prometheus));
  }
  if (config.openTelemetry) {
    const { useOpenTelemetry } = await import('@graphql-mesh/plugin-opentelemetry');
    plugins.push(useOpenTelemetry(config.openTelemetry));
  }

  if (config.rateLimiting) {
    const { default: useMeshRateLimit } = await import('@graphql-mesh/plugin-rate-limit');
    plugins.push(
      useMeshRateLimit({
        cache: ctx.cache,
        ...config.rateLimiting,
      }),
    );
  }

  return plugins;
}

export async function getCacheInstanceFromConfig(
  config: GatewayCLIBuiltinPluginConfig,
  ctx: Pick<GatewayConfigContext, 'logger' | 'pubsub'>,
): Promise<KeyValueCache> {
  if (config.cache && 'type' in config.cache) {
    switch (config.cache.type) {
      case 'redis': {
        const { default: RedisCache } = await import('@graphql-mesh/cache-redis');
        return new RedisCache({
          ...ctx,
          ...config.cache,
        });
      }
      case 'cfw-kv': {
        const { default: CloudflareKVCacheStorage } = await import('@graphql-mesh/cache-cfw-kv');
        return new CloudflareKVCacheStorage({
          ...ctx,
          ...config.cache,
        });
      }
    }
    if (config.cache.type !== 'localforage') {
      ctx.logger.warn('Unknown cache type, falling back to localforage', config.cache);
    }
    const { default: LocalforageCache } = await import('@graphql-mesh/cache-localforage');
    return new LocalforageCache({
      ...ctx,
      ...config.cache,
    });
  }
  if (config.cache) {
    return config.cache as KeyValueCache;
  }
  const { default: LocalforageCache } = await import('@graphql-mesh/cache-localforage');
  return new LocalforageCache();
}
