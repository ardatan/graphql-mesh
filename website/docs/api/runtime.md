---
id: runtime
title: Runtime
sidebar_label: Runtime
---

WIP!