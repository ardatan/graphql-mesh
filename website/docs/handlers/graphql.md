---
id: graphql
title: GraphQL
sidebar_label: GraphQL
---

This handler allows you to load remote GraphQL schemas and use it with schema-stitching, based on `graphql-tools-fork`.

To get started, install the handler library from NPM:

```
$ yarn add @graphql-mesh/graphql
```

Now, you can use it directly in your Mesh config file:

```yml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

## Config API Reference

{@import ../generated-markdown/GraphQLHandler.generated.md}
