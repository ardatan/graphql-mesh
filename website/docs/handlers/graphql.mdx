---
id: graphql
title: GraphQL
sidebar_label: GraphQL
---

![image](https://user-images.githubusercontent.com/20847995/79219047-333ccb00-7e5a-11ea-9fce-57ff137ba924.png)

This handler allows you to load remote GraphQL schemas and use them with schema-stitching based on `graphql-tools`.

To get started, install the handler library:

<PackageInstall packages="@graphql-mesh/graphql" />

Now, you can use it directly in your Mesh config file:

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
```

> You can check out our example that uses schema stitching with a PostgreSQL data source.
> [Click here to open the example on GitHub](https://github.com/Urigo/graphql-mesh/tree/master/examples/postgres-geodb)

<p>&nbsp;</p>

---

<p>&nbsp;</p>

## Dynamic Header Values (e.g., for Authorization)

Mesh can take dynamic values from the GraphQL Context or the environmental variables. For example, if you use `mesh dev` or `mesh start`, GraphQL Context will be the incoming HTTP request.

The expression inside dynamic values should be as in JS.

<p>&nbsp;</p>

### From Context (HTTP Header for `mesh dev` or `mesh start`)

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          # Please do not use capital letters while getting the headers
          Authorization: Bearer {context.headers['x-my-api-token']}
          # You can also access to the cookies like below;
          # Authorization: Bearer {context.cookies.myApiToken}
```

And for `mesh dev` or `mesh start`, you can pass the value using `x-my-graphql-api-token` HTTP header.

<p>&nbsp;</p>

### From Environmental Variable

`MY_API_TOKEN` is the name of the environmental variable you have the value.

```yaml
sources:
  - name: MyGraphQLApi
    handler:
      graphql:
        endpoint: http://my-service-url:3000/graphql
        operationHeaders:
          Authorization: Bearer {env.MY_API_TOKEN}
```

<p>&nbsp;</p>

---

<p>&nbsp;</p>

## Config API Reference

{@import ../generated-markdown/GraphQLHandlerHTTPConfiguration.generated.md}
or
{@import ../generated-markdown/GraphQLHandlerCodeFirstConfiguration.generated.md}
