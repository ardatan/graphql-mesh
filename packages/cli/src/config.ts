import { ConfigProcessOptions, processConfig } from '@graphql-mesh/config';
import { jsonSchema, YamlConfig } from '@graphql-mesh/types';
import { defaultImportFn, loadYaml } from '@graphql-mesh/utils';
import Ajv from 'ajv';
import { cosmiconfig, defaultLoaders } from 'cosmiconfig';
import path from 'path';

export function validateConfig(config: any): asserts config is YamlConfig.Config {
  const ajv = new Ajv({
    strict: false,
  } as any);
  jsonSchema.$schema = undefined;
  const isValid = ajv.validate(jsonSchema, config);
  if (!isValid) {
    console.warn(`Configuration is not valid:\n${ajv.errorsText()}`);
  }
}

export async function findAndParseConfig(options?: ConfigProcessOptions) {
  const { configName = 'mesh', dir: configDir = '', ...restOptions } = options || {};
  const dir = path.isAbsolute(configDir) ? configDir : path.join(process.cwd(), configDir);
  const explorer = cosmiconfig(configName, {
    loaders: {
      '.json': customLoader('json', options?.importFn),
      '.yaml': customLoader('yaml', options?.importFn),
      '.yml': customLoader('yaml', options?.importFn),
      '.js': customLoader('js', options?.importFn),
      '.ts': customLoader('js', options?.importFn),
      noExt: customLoader('yaml', options?.importFn),
    },
  });
  const results = await explorer.search(dir);

  if (!results) {
    throw new Error(`No ${configName} config file found in "${dir}"!`);
  }

  const config = results.config;
  validateConfig(config);
  return processConfig(config, { dir, ...restOptions });
}

function customLoader(ext: 'json' | 'yaml' | 'js', importFn = defaultImportFn) {
  function loader(filepath: string, content: string) {
    if (process.env) {
      content = content.replace(/\$\{(.*?)\}/g, (_, variable) => {
        let varName = variable;
        let defaultValue = '';

        if (variable.includes(':')) {
          const spl = variable.split(':');
          varName = spl.shift();
          defaultValue = spl.join(':');
        }

        return process.env[varName] || defaultValue;
      });
    }

    if (ext === 'json') {
      return defaultLoaders['.json'](filepath, content);
    }

    if (ext === 'yaml') {
      return loadYaml(filepath, content);
    }

    if (ext === 'js') {
      return importFn(filepath);
    }
  }

  return loader;
}
