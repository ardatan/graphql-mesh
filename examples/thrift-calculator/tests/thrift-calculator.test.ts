import { findAndParseConfig } from '@graphql-mesh/config';
import { getMesh } from '@graphql-mesh/runtime';
import { basename, join } from 'path';
import { introspectionFromSchema, lexicographicSortSchema } from 'graphql';
import { loadDocuments } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import thriftServer from '../src/main';
import { mkdirSync, writeFileSync } from 'fs';

const problematicModulePath = join(__dirname, '../../../node_modules/core-js/modules');
const emptyModuleContent = 'module.exports = {};';

// Fix core-js issue
mkdirSync(problematicModulePath, { recursive: true });
writeFileSync(join(problematicModulePath, './es.array.join.js'), emptyModuleContent);

const config$ = findAndParseConfig({
  dir: join(__dirname, '..'),
});
const mesh$ = config$.then(config => getMesh(config));
jest.setTimeout(30000);

describe('Thrift Calculator', () => {
  it('should generate correct schema', async () => {
    const { schema } = await mesh$;
    expect(
      introspectionFromSchema(lexicographicSortSchema(schema), {
        descriptions: false,
      })
    ).toMatchSnapshot('thrift-calculator-schema');
  });
  it('should give correct response for example queries', async () => {
    const {
      config: {
        serve: { exampleQuery },
      },
    } = await config$;
    const sources = await loadDocuments(join(__dirname, '..', exampleQuery), {
      loaders: [new GraphQLFileLoader()],
    });
    const { execute } = await mesh$;
    for (const source of sources) {
      const result = await execute(source.document, {});
      expect(result.errors).toBeFalsy();
      expect(result).toMatchSnapshot(basename(source.location) + '-thrift-calculator-result');
    }
  });
  afterAll(() => {
    mesh$.then(mesh => mesh.destroy());
    thriftServer.close();
  });
});
