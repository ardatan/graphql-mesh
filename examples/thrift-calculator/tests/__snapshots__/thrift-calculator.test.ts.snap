// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thrift Calculator executes Add.query.graphql query: Add.query.graphql 1`] = `
{
  "data": {
    "add": 5,
  },
  "errors": undefined,
  "extensions": undefined,
}
`;

exports[`Thrift Calculator generates the schema correctly: schema 1`] = `
"schema @transport(subgraph: "calculator", kind: "thrift", location: "http://localhost:9876/thrift", headers: {}, options: {clientAnnotations: {serviceName: "Calculator", annotations: {}, methodNames: ["add", "subtract"], methodAnnotations: {add: {annotations: {}, fieldAnnotations: {}}, subtract: {annotations: {}, fieldAnnotations: {}}}, methodParameters: {add: 2, subtract: 3}}, topTypeMap: {AddRequest: {id: 0, name: "AddRequest", type: 12, fields: {left: {type: 8, id: 1}, right: {type: 8, id: 2}}}}}) {
  query: Query
}

type Query {
  add(request: AddRequestInput): Int @fieldTypeMap(request: {name: "AddRequest", type: "ref", id: 1}) @resolver(subgraph: "calculator", operation: "query add($request: AddRequestInput) { add(request: $request) }") @source(subgraph: "calculator", name: "add", type: "Int")
  subtract(left: Int, right: Int): Int @fieldTypeMap(left: {type: 8, id: 1}, right: {type: 8, id: 2}) @resolver(subgraph: "calculator", operation: "query subtract($left: Int, $right: Int) { subtract(left: $left, right: $right) }") @source(subgraph: "calculator", name: "subtract", type: "Int")
}

input AddRequestInput @source(subgraph: "calculator", name: "AddRequestInput") {
  left: Int! @source(subgraph: "calculator", name: "left", type: "Int!")
  right: Int! @source(subgraph: "calculator", name: "right", type: "Int!")
}"
`;
